### PR TITLE
sql: fix cast from int to char by interpreting integer as ASCII byte

### LINKED ISSUE #70771: sql: cast from integer to "char" should interpret the integer as an ASCII byte
In Postgres:

```
marcus=# SELECT 97::"char";
 char
------
 a
(1 row)

marcus=# SELECT 128::"char";
ERROR:  22003: "char" out of range
```

In CockroachDB v21.2.x:

```
demo@127.0.0.1:26257/defaultdb> SELECT 97::"char";
  char
-------
    9
(1 row)

demo@127.0.0.1:26257/defaultdb> SELECT 128::"char";
  char
-------
   1
(1 row)
```

### PR DESCRIPTION
**Pull request by techytoes**
Fixes: #70771 

Release note (sql change): Earlier the behavior of casting an int
to "char" was similar to bpchar where we returned only
the first digit of the integer. Now, we are interpreting
int cast to "char" as an ASCII byte to align the overall
behavior more with PostgreSQL.

Release note (sql change): A parameter of type "char" can now be used as
a parameter in a prepared statement.

### CODE REVIEW

**Review by mgartner — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Can you add some logic and tests for integers that exceed the maximum standard ASCII value (127)? We should error with a message similar to Postgres:

```
marcus=# select 127::"char";
 char
------
 \x7F
(1 row)

marcus=# select 128::"char";
ERROR:  22003: "char" out of range
```

We should also have some tests for when the integer is negative. It looks like Postgres casts negative integers > -129 to a whitespace character, and errors when casts negative integers <= -129:

```
marcus=# select i, i::"char", length(i::"char") from (values (-1), (-25), (-128)) v(i);
  i   | i | length
------+---+--------
   -1 |   |      1
  -25 |   |      1
 -128 |   |      1
(3 rows)

marcus=# select (-129)::"char";
ERROR:  22003: "char" out of range
```

**Review by mgartner — COMMENTED** [review]
I was wrong earlier regarding negative integers:

```
marcus=# select (-1)::"char" = ' '::"char";
 ?column?
----------
 f
(1 row)
```

**Comment by mgartner** [comment]
It looks like there is a special case for `0` too: 

```
marcus=# select i, i::"char", length(i::"char"), i::"char" = ' '::"char" from (values (97), (127), (0), (-1), (-127), (-128)) v(i);
  i   |  i   | length | ?column?
------+------+--------+----------
   97 | a    |      1 | f
  127 | \x7F |      1 | f
    0 |      |      0 | f
   -1 |      |      1 | f
 -127 |      |      1 | f
 -128 |      |      1 | f
(6 rows)
```

**Review by mgartner — COMMENTED** [review]
One more nitpick: The `"char"` type is distinct from `CHAR` (which is an alias for `CHAR(1)`). This is an annoying detail, but I think it's important to use "char" in the PR title/description and the commit message.

**Comment by mgartner** [comment]
Postgres's code for casting an integer to a `"char"` is here: https://github.com/postgres/postgres/blob/b484ddf4d2eb81736512efa35ed3e5d2a72993d8/src/backend/utils/adt/char.c#L161-L172

Where `SCHAR_MIN` is -128 and `SCHAR_MAX` is 127.

**Comment by techytoes** [comment]
@mgartner yes, I have made relevant changes to handle edge cases. I am little unsure about the PostgreSQL interpretation of `0::"char"` though.
```
postgres=# select 0::"char";
 char
------

(1 row)

postgres=# select 32::"char";
 char
------

(1 row)
```
both 0 and 32 returns the same result. Let me know your thoughts regarding this? Thanks!

**Comment by rafiss** [comment]
they actually aren't the same, but it's hard to tell because it's whitespace:
```
rafiss@127:postgres> select i, i::"char", length(i::"char"), i::"char" = ' '::"char" from (values (97), (127), (0), (32)) v(i);
+-----+-----+----------+------------+
| i   | i   | length   | ?column?   |
|-----+-----+----------+------------|
| 97  | a   | 1        | False      |
| 127 |      | 1        | False      |
| 0   |     | 0        | False      |
| 32  |     | 1        | True       |
+-----+-----+----------+------------+
```

i'd recommend adding this `select i, i::"char", length(i::"char") from (values (0), (32)) v(i);` query as a test case in `pkg/sql/logictest/testdata/logic_test/cast` so that this is easier to assert on. (and also include more values in that `values` clause too.)

although @mgartner i get different behavior than you for negative numbers:
```
rafiss@127:postgres> select i, i::"char", length(i::"char"), i::"char" = ' '::"char" from (values (-1)) v(i);
'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

**Comment by rafiss** [comment]
update: the difference with negative numbers was client side. i was testing with `pgcli`. if i test with `psql` my results match @mgartner 

so let's add this test case to a logic test, as well as some out of range cases

```
rafiss=# select i, i::"char", length(i::"char"), i::"char" = ' '::"char" from (values (32), (97), (127), (0), (-1), (-127), (-128)) v(i);
  i   |  i   | length | ?column?
------+------+--------+----------
   32 |      |      1 | t
   97 | a    |      1 | f
  127 | \x7F |      1 | f
    0 |      |      0 | f
   -1 |      |      1 | f
 -127 |      |      1 | f
 -128 |      |      1 | f
(7 rows)

rafiss=# select 128::"char";
ERROR:  "char" out of range

rafiss=# select (-129)::"char";
ERROR:  "char" out of range
```

**Comment by mgartner** [comment]
We'll probably also want a test that ensures that the value of chars cast from negative integers have the correct underlying value, since it can't be seen in the output. For example:

```
marcus=# select (-127)::"char" = (-126)::"char";
 ?column?
----------
 f
(1 row)
```

**Comment by rafiss** [comment]
here's a diff for a test that we can add that will make sure we handle negative numbers correctly
```
diff --git a/pkg/cmd/generate-binary/main.go b/pkg/cmd/generate-binary/main.go
index ed28fb6549..07f3de744d 100644
--- a/pkg/cmd/generate-binary/main.go
+++ b/pkg/cmd/generate-binary/main.go
@@ -547,4 +547,15 @@ var inputs = map[string][]string{
 		`1::int8,(2::int8,3::int8)`,
 		`1::int8,('hi'::TEXT,3::int2)`,
 	},
+
+	`%s::"char"`: {
+		`(-128)`,
+		`(-32)`,
+		`(-1)`,
+		`0`,
+		`1`,
+		`32`,
+		`97`,
+		`127`,
+	},
 }
diff --git a/pkg/sql/pgwire/testdata/encodings.json b/pkg/sql/pgwire/testdata/encodings.json
index 9c3f549c69..af9fb803d0 100644
--- a/pkg/sql/pgwire/testdata/encodings.json
+++ b/pkg/sql/pgwire/testdata/encodings.json
@@ -41,6 +41,62 @@
 		"TextAsBinary": [123, 125],
 		"Binary": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 23]
 	},
+	{
+		"SQL": "(-128)::\"char\"",
+		"Oid": 18,
+		"Text": "\x80",
+		"TextAsBinary": [128],
+		"Binary": [128]
+	},
+	{
+		"SQL": "(-32)::\"char\"",
+		"Oid": 18,
+		"Text": "\xe0",
+		"TextAsBinary": [224],
+		"Binary": [224]
+	},
+	{
+		"SQL": "(-1)::\"char\"",
+		"Oid": 18,
+		"Text": "\xff",
+		"TextAsBinary": [255],
+		"Binary": [255]
+	},
+	{
+		"SQL": "0::\"char\"",
+		"Oid": 18,
+		"Text": "",
+		"TextAsBinary": [],
+		"Binary": [0]
+	},
+	{
+		"SQL": "1::\"char\"",
+		"Oid": 18,
+		"Text": "\x01",
+		"TextAsBinary": [1],
+		"Binary": [1]
+	},
+	{
+		"SQL": "32::\"char\"",
+		"Oid": 18,
+		"Text": " ",
+		"TextAsBinary": [32],
+		"Binary": [32]
+	},
+	{
+		"SQL": "97::\"char\"",
+		"Oid": 18,
+		"Text": "a",
+		"TextAsBinary": [97],
+		"Binary": [97]
+	},
+	{
+		"SQL": "127::\"char\"",
+		"Oid": 18,
+		"Text": "\u007f",
+		"TextAsBinary": [127],
+		"Binary": [127]
+	},
 	{
 		"SQL": "'hello'::char(8)",
 		"Oid": 1042,
```

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @techytoes)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAn_Pq-tdFYVY6UYU0:-MlAn_Pq-tdFYVY6UYU1:b-amhy1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/logictest/testdata/logic_test/cast#L87)):*
> ```
> 
> query
> SELECT i, i::"char"
> ```

can you make it `SELECT i, i::"char", length(i::"char")
___
*[pkg/sql/sem/tree/casts.go, line 995 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAnk2BB3EmQ-X4aj-I:-MlAnk2BB3EmQ-X4aj-J:b5f2hpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/casts.go#L995)):*
> ```Go
> 			if typ.Oid() == oid.T_char {
> 				if *t > math.MaxInt8 || *t < math.MinInt8 {
> 					return nil, ErrCharOutOfRange
> ```



although this error makes sense, i checked with postgres and it returns "numeric_value_out_of_range" (22003) for this
___
*[pkg/sql/sem/tree/casts.go, line 997 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAnm1l933HOhFh5Y8c:-MlAnm1m6ljRhLGDJgBo:b-ehybcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/casts.go#L997)):*
> ```Go
> 					return nil, ErrCharOutOfRange
> 				} else if *t <= 0 {
> 					s = ""
> ```

this should be:
```
else if *t == 0 {
  s = ""
} else
  s = string(rune(*t))
}
```
___
*[pkg/sql/sem/tree/eval.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAo1Jw2rBFuFAiYWoo:-MlAo1Jw2rBFuFAiYWop:b7i3574) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/eval.go#L68)):*
> ```Go
> 
> 	// ErrCharOutOfRangeis reported when int cast to ASCII byte overflows.
> 	ErrCharOutOfRange = pgerror.New(pgcode.InvalidCharacterValueForCast, "\"char\" out of range")
> ```

see above comment,  we don't need this error

**Review by techytoes — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAn_Pq-tdFYVY6UYU0:-MlB4rrY19bZVm81q_--:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/logictest/testdata/logic_test/cast#L87)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you make it `SELECT i, i::"char", length(i::"char")
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/casts.go, line 995 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAnk2BB3EmQ-X4aj-I:-MlB4quX7OjTz-aM8HbA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/casts.go#L995)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

although this error makes sense, i checked with postgres and it returns "numeric_value_out_of_range" (22003) for this
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/casts.go, line 997 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAnm1l933HOhFh5Y8c:-MlB225kEW7Dtum70TTV:b-puqugl) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/casts.go#L997)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this should be:
```
else if *t == 0 {
  s = ""
} else
  s = string(rune(*t))
}
```
</blockquote></details>

We might also need to add a separate condition for negative numbers(>math.MinInt8). Seems like Postgres returns a spaced string for these numbers.

```
postgres=# select (-10)::"char", length((-10)::"char");
 char | length
------+--------
      |      1
(1 row)

```

```
demo@127.0.0.1:26257/movr> select (-1)::"char", length((-1)::"char");
  char | length
-------+---------
  �    |      1
(1 row)
```
___
*[pkg/sql/sem/tree/eval.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAo1Jw2rBFuFAiYWoo:-MlB4oOj-dcuX2QOFz0F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/eval.go#L68)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

see above comment,  we don't need this error
</blockquote></details>

Done.

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[pkg/sql/sem/tree/casts.go, line 997 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAnm1l933HOhFh5Y8c:-MlB5XvE3EeWQfuD7gHC:bdnti3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/casts.go#L997)):*
<details><summary><i>Previously, techytoes (Rupesh Harode) wrote…</i></summary><blockquote>

We might also need to add a separate condition for negative numbers(>math.MinInt8). Seems like Postgres returns a spaced string for these numbers.

```
postgres=# select (-10)::"char", length((-10)::"char");
 char | length
------+--------
      |      1
(1 row)

```

```
demo@127.0.0.1:26257/movr> select (-1)::"char", length((-1)::"char");
  char | length
-------+---------
  �    |      1
(1 row)
```
</blockquote></details>

i don't think that's correct. see @mgartner's comments:

we can see exactly what postgres is doing: https://github.com/cockroachdb/cockroach/pull/70942#issuecomment-931595085

and also, `i::"char" = ' '::"char"` returns false for negative numbers. https://github.com/cockroachdb/cockroach/pull/70942#issuecomment-931571510


the difference you are seeing is likely just because of client-side rendering differences with `psql` versus `cockroach sql`.

it's not easy to tell them apart, but there's a difference between a non-printable character versus a whitespace character. the best way we can actually see the difference is with the encoding_test that i asked you to add. it shows that these negative numbers are not whitespace. for example, for the `(-1)::"char"`test case, it shows that the text representation is `\xFF`

**Comment by techytoes** [comment]
@rafiss got it. I somehow missed that comment. Thanks for the help! 🙇🏼

**Review by rafiss — CHANGES_REQUESTED** [review]
looks like the encoding test is going to have some trouble with these non-printable characters. i'll think of a way to fix the encoding test, but in the mean time left a few more comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-:-MlBlDm--OollswoMNQn:b-uqho61)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @techytoes)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlBkyrh3uitMHBkgPN-:-MlBkyri28SYKsm_iHg4:b-521fmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ce2d01feacb511e4c6fdb49fcf18cdaec5f7c78/pkg/sql/logictest/testdata/logic_test/cast#L86)):*
> ```
> 1  bar  1  foo
> 
> SELECT i, i::"char", length(i::"char")
> ```

you need `query ITI` on the line before the SELECT
___
*[pkg/sql/sem/tree/eval.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlAo1Jw2rBFuFAiYWoo:-MlBl45pAD_XaJe4ACjf:bvkk4vg) ([raw file](https://github.com/cockroachdb/cockroach/blob/587dff0573346df129b7166c327544237744780d/pkg/sql/sem/tree/eval.go#L68)):*
<details><summary><i>Previously, techytoes (Rupesh Harode) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

on second thought, could you keep this, but make it defined as
```
	// errCharOutOfRange is reported when int cast to ASCII byte overflows.
	errCharOutOfRange = pgerror.New(pgcode.NumericValueOutOfRange, "\"char\" out of range")
```

that way we still get the right error message

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @techytoes)

___
*[pkg/sql/sem/tree/casts.go, line 999 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70942#-MlCpHirCeXr9vsu8lme:-MlCpHiu89jZr4owO_pZ:bpwq0nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ce2d01feacb511e4c6fdb49fcf18cdaec5f7c78/pkg/sql/sem/tree/casts.go#L999)):*
> ```Go
> 					s = ""
> 				} else {
> 					s = string(rune(*t))
> ```

i looked into this and it turns out it's not correct to use `rune` here. it needs to be:
```
string([]byte{byte(*t)})
```

**Comment by rafiss** [comment]
After looking into the encoding tests, I added a new revision to this PR myself. It turned out to be quite tricky, and because of the non-printable chars, I couldn't think of an easy way to send the diff to you.

I also had to make a change to pgwire so that the "char" type can be properly encoded, so it can be tested by the encoding_test. I cleaned up the logic test as well while I was at it.

I'll go ahead and merge this after the CI is passing.

Thanks for making all the changes here!

**Review by rafiss — APPROVED** [review]

File: pkg/sql/sem/tree/eval.go [hunk 1]
```diff
@@ -63,6 +63,8 @@ var (
 	// ErrFloatOutOfRange is reported when float arithmetic overflows.
 	ErrFloatOutOfRange = pgerror.New(pgcode.NumericValueOutOfRange, "float out of range")
 	errDecOutOfRange   = pgerror.New(pgcode.NumericValueOutOfRange, "decimal out of range")
+	// errCharOutOfRange is reported when int cast to ASCII byte overflows.
+	errCharOutOfRange = pgerror.New(pgcode.NumericValueOutOfRange, "\"char\" out of range")
```
Comment (otan) [inline pkg/sql/sem/tree/eval.go:67]: optional nit:
```
`"char" out of range`
```

**Review by otan — APPROVED** [review]

**Pull request merged**